### ISSUE TITLE
Recursive default iterators don't resolve unless they use .these()

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#5216]
### Summary of Problem

When a default (``these()``) iterator calls itself recursively, implicitly, the compiler fails to resolve it.  If the recursive call is made explicitly then things work as expected.

### Steps to Reproduce

**Source Code:**

(taken from test/functions/iterators/recursive/recursive-iter.chpl added in PR #5215)

```chapel
class Tree {
  var data: int;
  var left: Tree;
  var right: Tree;

  iter these(): int {
    if left then
      for e in left do yield e;
    yield data;
    if right then
      for e in right do yield e;
  }
}

var tree = new Tree(40, new Tree(20, new Tree(10), new Tree(30)), new
                    Tree(60, new Tree(50), new Tree(70)));

for e in tree do
  writeln(e);
```

Changing the `in left` and `in right` invocations to `in left.these()` and `in right.these()` causes this to work as intended.

**Compile command:**
`chpl recursive-iter.chpl`

**Execution command:**
`./a.out`

### Configuration Information

- Output of `chpl --version`:  1.14.0.0bea73a
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:  N/A
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:  N/A
- (For Cray systems only) Output of `module list`:  N/A

**Comment by cassella** [comment]
The future added for this in #5215 was retired in #13574, apparently as a result of making the tree's nodes owned nilable, or of iterating over `left!` and `right!` instead of `left` and `right`.

The code as written above still hits an internal error,

```
fortytwo@magrathea:~/src/chapel (master)$ chpl foo.chpl
foo.chpl:7: internal error: Unhandled type in blankIntentForType() [passes/resolveIntents.cpp:132]
Note: This source location is a guess.
```
or
```
fortytwo@magrathea:~/src/chapel (master)$ chpl --no-devel foo.chpl
foo.chpl:7: internal error: PAS-RES-NTS-0132 chpl version 1.21.0 pre-release (7cecda3)
```

**Comment by bradcray** [comment]
@cassella: Do you want to put together a PR that reverts the code to its original form and restores the future?  @vasslitvinov was there a good reason to rewrite the future such that it was retired?

**Comment by cassella** [comment]
As written, the original code shouldn't be legal any more as it requires default-initializing the non-nilable left and right members of the leaf nodes (IIUC).  The new form looks like a faithful adaptation of the intent of the old code in a nilable-classes world.  And it does what the title of the issue says doesn't work, which seems worth locking in.

So I'd guess it's worth keeping both versions.  I don't know what the `.good` file should be for the original code if it should fail due to nilability.  I tried figuring it out by removing the iterator, but

```chapel
class Tree {
  var data: int;
  var left: Tree;
  var right: Tree;
}

var tree = new Tree(40, new Tree(20, new Tree(10), new Tree(30)), new
                    Tree(60, new Tree(50), new Tree(70)));

writeln(tree);
```

```
$CHPL_HOME/modules/internal/ChapelIO.chpl:280: internal error: Unhandled type in constIntentForType() [passes/resolveIntents.cpp:63]
```


These internal errors seem related to each other, but not like the "failure to resolve" this issue was opened for.  Perhaps this error didn't happen in master at the point in time when the test was updated for no-legacy-classes?

**Comment by cassella** [comment]
I think these errors are unrelated to the failure to resolve a recursive default iterator not invoked via `.these()`:  They're not "failure to resolve" errors, and one of them happens without the iterator.

I'd be more inclined to open a new issue than to attach futures for these to this issue.  But compiler/language changes to address objects containing nonnilable class may make it moot.

Bonus: the no-iterator version above, when the call to writeln() is also removed, fails to compile the generated source:

```
[Compiler output was as follows:]
In file included from /tmp/chpl-fortytwo-23021.deleteme/_main.c:2:0:
/tmp/chpl-fortytwo-23021.deleteme/chpl__header.h:2801:1: error: unknown type name ‘_TypeDefaultT’
 _TypeDefaultT left_chpl;
 ^
/tmp/chpl-fortytwo-23021.deleteme/chpl__header.h:2802:1: error: unknown type name ‘_TypeDefaultT’
 _TypeDefaultT right_chpl;
 ^
/tmp/chpl-fortytwo-23021.deleteme/chpl__header.h:5742:1: error: unknown type name ‘_TypeDefaultT’
 _TypeDefaultT left_chpl,
 ^
/tmp/chpl-fortytwo-23021.deleteme/chpl__header.h:5743:1: error: unknown type name ‘_TypeDefaultT’
 _TypeDefaultT right_chpl,
 ^
/tmp/chpl-fortytwo-23021.deleteme/chpl__header.h:5756:1: error: unknown type name ‘_TypeDefaultT’
 _TypeDefaultT left_chpl,
 ^
/tmp/chpl-fortytwo-23021.deleteme/chpl__header.h:5757:1: error: unknown type name ‘_TypeDefaultT’
 _TypeDefaultT right_chpl,
 ^
In file included from /tmp/chpl-fortytwo-23021.deleteme/_main.c:48:0:
/tmp/chpl-fortytwo-23021.deleteme/recursive-iter-classic-no-iter-no-writeln.c: In function ‘chpl__init_recursive_HYPHEN_iter_HYPHEN_classic_HYPHEN_no_HYPHEN_iter_HYPHEN_no_HYPHEN_writeln’:
/tmp/chpl-fortytwo-23021.deleteme/recursive-iter-classic-no-iter-no-writeln.c:16:17: error: implicit declaration of function ‘_new_chpl21’ [-Werror=implicit-function-declaration]
 new_temp_chpl = _new_chpl21(INT64(40), _typeDefaultT, _typeDefaultT, INT64(7), INT32(60));
                 ^
...
```

Maybe these new errors are more about the class itself being recusrive, in that it contains generic-ish same-class elements?  I'd probably put the new futures in `test/classes`, but I don't know which of `test/classes/{delete-free,initializers,generic,nilability}` would be the appropriate home.

**Comment by cassella** [comment]
I opened #14378 #14379 and #14380 for the problems I've run into.  In them, I was able to simplify out the recursion and the iteration.  The common thread there seems to be a class or record having an `anymanaged` member.  So I think they're all unrelated to this issue.

I think the `recursive-iter.chpl` test passing in its current state means that the problem described in this issue is resolved.  Or at the very least, it leaves it without a test case.
